lib/analyse.ml

```diff
@@ -7,12 +7,11 @@ let ( / ) = Filename.concat
 let ( >>!= ) = Lwt_result.bind
 
 let read_file ~max_len path =
-  let ch = open_in path in
-  Fun.protect ~finally:(fun () -> close_in ch)
-    (fun () ->
-       let len = in_channel_length ch in
-       if len <= max_len then really_input_string ch len
-       else Fmt.failwith "File %S too big (%d bytes)" path len
+  Lwt_io.with_file ~mode:Lwt_io.input path
+    (fun ch ->
+       Lwt_io.length ch >>= fun len ->
+       if len <= Int64.of_int max_len then Lwt_io.read ~count:max_len ch
```

Comment:
Rather surprisingly, `Lwt_io.read ~count` does not read `count` bytes. You need to use `Lwt_io.read_into_exactly` instead.

> read ~count ic reads at most ~count characters from ic

However, if you don't specify `count` then it *does* read all the bytes!

Reply:
Does it really matter if it doesn't? I believe the semantics is "read as much as possible to a maximum length of `count`" and we already checked its length just before. Unless the file changes between the two calls I believe this should return a string exactly of size `len`.

Reply:
It does matter. I already had this bug in the solver code :-( If the OS is busy, it can unpredictably return a short read.

Reply:
My bad. Given that `Lwt_io` is supposed to be somewhat higher level than `Lwt_unix`, I thought this function would read repetitively until the end of the file or `count`, not just read once.

Fixed now.

Reply:
@aantron : the `Lwt_io.read` API has confused both of us (see above). Any ideas how this can be made safer?

Reply:
I suppose we could quietly turn `read ~count` into `read_into_exactly` for strings. This would be still be consistent with the docs, which describe the behavior you observed. It looks like `read`'s two behaviors are really attempts to expose the implementation detail that with `~count`, `read` tries to do exactly one read, and without it, `read` reads in a loop. This is the cause of the confusion, and it was probably a mistake to overload `read` to have these two different behaviors. We should just make `read` always read in a loop, just with a limit or not.

Note that `read` can still return less than `count` bytes in case end of input is reached, which is probably the most common reason why this would happen. That wasn't what was biting you in this issue, was it?

Reply:
In the bug I had, it was reading less than `count` when the system was under load IIRC. Perhaps the argument could be called `limit` when it's just a bound? e.g.

- `read_string : ?count:int -> t -> string` (read exactly `count` if given, or the whole file if not)
- `read_single : limit:int -> t -> string`

Though I'm not sure if we actually need the second one. I bet most people who want a single read are using `read_into` already. Not sure about the name either. `read_exactly` is clear when `count` is provided, but not for reading the whole file. Could be two functions I suppose.

In any case, it would be good to clarify in the docs that end-of-file isn't the only reason why you may get a short read.

Reply:
I updated the docs for now in https://github.com/ocsigen/lwt/commit/2daa73b2aec332c0c929b73161fb261a8c798e6e. If this bites someone again, I'd probably choose to make `read ~count` read in a loop unless there is end of input or an error.